scimma_admin/hopskotch_auth/api_views.py

```diff
@@ -48,6 +64,137 @@ def get(self, request):
+                                                           s_nonce=ex.s_nonce())
+    s.set_client_first(ex.client_first)
+    s.get_server_first()  # waste of time, but scramp requires this to be called
+    # if we reach this point, we are ready to process the second half of the exchange
+    s.set_client_final(client_final)
+    # if scramp hasn't objected, the authentication has now succeeded
+    return (ex,s)
+
+def parse_list_header(header: str):
+    return [v[1:-1] if v[0] == v[-1] == '"' else v for v in parse_http_list(header)]
+
+def parse_dict_header(header: str):
+    def unquote(v: str):
+        return v[1:-1] if v[0] == v[-1] == '"' else v
+    d = dict()
+    for item in parse_list_header(header):
+        if '=' in item:
+            k, v = item.split('=', 1)
+            d[k] = unquote(v)
+        else:
+            d[k] = None
+    return d
+
+class ScramState(object):
+    def __init__(self, mech, sid, s):
+        self.mech = mech
+        self.sid = sid
+        self.s = s
+
+class ScramAuthentication(BaseAuthentication):
+    def authenticate(self, request):
+        auth_header = get_authorization_header(request)
+        if not auth_header or len(auth_header)==0:
+            return None
+        try:
+            auth_header=auth_header.decode("utf-8")
+        except:
+            raise AuthenticationFailed("Malformed authentication header")
+
+        if not auth_header.upper().startswith("SCRAM-"):
+            return None
+        m = re.fullmatch("(SCRAM-[A-Z0-9-]+) *([^ ].*)", auth_header, flags=re.IGNORECASE)
+        if not m:
+            raise AuthenticationFailed("Malformed SCRAM authentication header")
+        scram_mech=m.group(1).upper()
+        auth_data = parse_dict_header(m.group(2))
+        if "data" in auth_data and "sid" in auth_data:
+            # If we have both of these we are in the final phase of the SCRAM handshake
+            sid = auth_data.get("sid")
+            data = auth_data.get("data")
+            if not sid or not data:
+                raise AuthenticationFailed("Malformed SCRAM authentication header")
+            client_final=base64.b64decode(data).decode("utf-8")
+            ex,s = do_scram_final(client_final, sid)
+            request.META["scram_state"]=ScramState(scram_mech, sid, s)
+            return (ex.cred.owner, ex.cred)
+        # Otherwise, SCRAM has not yet succeeded
+        return None
+
+    def authenticate_header(self, request):
+        auth_header = get_authorization_header(request)
+        if not auth_header or len(auth_header)==0:
+            return "SCRAM-SHA-512"
+        try:
+            auth_header=auth_header.decode("utf-8")
+        except:
+            raise AuthenticationFailed("Malformed SCRAM authentication header")
+        if auth_header.upper().startswith("SCRAM-"):
+            m = re.fullmatch("(SCRAM-[A-Z0-9-]+) *([^ ].*)", auth_header, flags=re.IGNORECASE)
```

Comment:
This regex search and match manipulation does not appear to be in a try/except block. Is it okay if this errors? Or are you sure will not?

Reply:
The official python documentation is lamentably unclear on whether exceptions can be raised, and if so what kinds and under what conditions. The obvious failure mode is a lack of a match, which should be handled by the subsequent code, but if there are possible exceptions I do want to handle those as well.